### PR TITLE
thrift: supported setting max requests for per downstream connection

### PR DESCRIPTION
**Pull request by wbpcode**
Signed-off-by: wbpcode <comems@msn.com>

Commit Message: thrift: supported setting max requests for per downstream connection
Additional Description:

Supported setting max requests for per downstream connection. By setting max_requests_per_connection, Envoy can actively disconnect the thrift client after reaching a certain number of requests.
Due to the limitations of the thrift protocol itself, we cannot achieve a clean disconnection. 

Check #14560 get more information.

Risk Level: Normal
Testing: Added
Docs Changes: N/A
Release Notes: Added

### CODE REVIEW

**Comment by zyfjeff** [comment]
I feel that this requirement is not universal. Why not do this by removing the listener from the control plane? @wbpcode

**Comment by wbpcode** [comment]
> I feel that this requirement is not universal. Why not do this by removing the listener from the control plane? @wbpcode

That's also an option. However, if there are multiple dubbo services in the cluster, although some of them have generated the corresponding listener for it, such as 1.2.3.4:10000, there may still be traffic accessing other dubbo services that need to be proxied through 0.0.0.0:10000.

In addition, we have a similar configuration for upstream connections, and I personally don't see any harm in making the same restrictions for downstream connections. 😄

**Comment by wbpcode** [comment]
@rgs1 Would you mind helping to give this pr a review or some suggestions? 😃

File: source/extensions/filters/network/thrift_proxy/config.h
Comment (rgs1) [inline source/extensions/filters/network/thrift_proxy/config.h:102]: nit:

```
uint64_t max_requests_per_connection_{};
```

initializing this to zero and skipping the limit check when zero is probably easier (and more common across the code base).

File: source/extensions/filters/network/thrift_proxy/conn_manager.h
Comment (rgs1) [inline source/extensions/filters/network/thrift_proxy/conn_manager.h:278]: I _think_ we can drop the 2nd sentence since it's a common assumption that conn managers are per thread.

File: source/extensions/filters/network/thrift_proxy/conn_manager.cc
Comment (rgs1) [inline source/extensions/filters/network/thrift_proxy/conn_manager.cc:419]: can we add a stat for this and increment it here?

File: test/extensions/filters/network/thrift_proxy/conn_manager_test.cc
Comment (rgs1) [inline test/extensions/filters/network/thrift_proxy/conn_manager_test.cc:1106]: Can we also test:

* limit set but not reached
* limit set, reached, cleared and next request goes through

And also test the suggested stat from above is properly incremented in each case.

**Review by rgs1 — CHANGES_REQUESTED** [review]
Thanks! We are going to use this too, so excited to see this feature. 

Could you also add a comment about this new feature in `docs/root/configuration/listeners/network_filters/thrift_proxy_filter.rst`?

cc: @fishcakez to get additional feedback

File: source/extensions/filters/network/thrift_proxy/config.h
Comment (wbpcode) [reply source/extensions/filters/network/thrift_proxy/config.h:102]: Get it.

File: source/extensions/filters/network/thrift_proxy/conn_manager.h
Comment (wbpcode) [reply source/extensions/filters/network/thrift_proxy/conn_manager.h:278]: :ok_hand:

File: source/extensions/filters/network/thrift_proxy/conn_manager.cc
Comment (wbpcode) [reply source/extensions/filters/network/thrift_proxy/conn_manager.cc:419]: Get.

File: source/extensions/filters/network/thrift_proxy/config.h
Comment (rgs1) [inline source/extensions/filters/network/thrift_proxy/config.h:3]: not needed anymore i think

File: source/extensions/filters/network/thrift_proxy/conn_manager.h
Comment (rgs1) [inline source/extensions/filters/network/thrift_proxy/conn_manager.h:278]: Can we do this the other way around? Having a counter of seen requests and a the limit in a separate var? E.g.:

```
  uint64_t requests_{};
  uint64_t max_requests_{}; 
```

If max_requests_ is 0, there are no limits. I rather avoid depending on `std::numeric_limits<uint64_t>::max()`.

File: test/extensions/filters/network/thrift_proxy/conn_manager_test.cc
Comment (rgs1) [inline test/extensions/filters/network/thrift_proxy/conn_manager_test.cc:1244]: Why is this loop needed?

**Review by rgs1 — CHANGES_REQUESTED** [review]
Thanks for the updates, left a few more comments.

File: source/extensions/filters/network/thrift_proxy/config.h
Comment (wbpcode) [reply source/extensions/filters/network/thrift_proxy/config.h:3]: 👍 yes, it is not needed. I just forget to remove it.

File: test/extensions/filters/network/thrift_proxy/conn_manager_test.cc
Comment (wbpcode) [reply test/extensions/filters/network/thrift_proxy/conn_manager_test.cc:1244]: In order to simulate the situation of multiple disconnections, observe whether the stat value grows correctly.

File: source/extensions/filters/network/thrift_proxy/conn_manager.h
Comment (wbpcode) [reply source/extensions/filters/network/thrift_proxy/conn_manager.h:278]: I have no objection. The reason for choosing this method is to reduce one `if` judgment. If you use this method, you need to determine whether max_requests is 0, and then determine whether the accumulated request value is greater than max_requests.

Comment (wbpcode) [reply source/extensions/filters/network/thrift_proxy/conn_manager.h:278]: But it should be ok.

File: source/extensions/filters/network/thrift_proxy/conn_manager.cc
Comment (rgs1) [inline source/extensions/filters/network/thrift_proxy/conn_manager.cc:417]: this should actually be:

```
parent_.accumulated_requests_ > parent_.config_.maxRequestsPerConnection()
```

if I set this to 50 requests... I want 50 requests to go through.

File: test/extensions/filters/network/thrift_proxy/conn_manager_test.cc
Comment (rgs1) [inline test/extensions/filters/network/thrift_proxy/conn_manager_test.cc:124]: nit: `uint32_t sendRequests(uint32_t count) {`

File: test/extensions/filters/network/thrift_proxy/conn_manager_test.cc
Comment (rgs1) [inline test/extensions/filters/network/thrift_proxy/conn_manager_test.cc:1122]: see my comment above, this is counter intuitive... 50 requests should be able to go through if 50 is the configured limit.

File: test/extensions/filters/network/thrift_proxy/conn_manager_test.cc
Comment (rgs1) [inline test/extensions/filters/network/thrift_proxy/conn_manager_test.cc:1201]: should we check here if `cx_destroy_local_with_active_rq` was incremented? I think it probably is incremented when a close happens after max requests is reached.

**Review by rgs1 — COMMENTED** [review]
This looks great, thanks! I have one final nit and one question about whether we should check if `cx_destroy_local_with_active_rq` is incremented after a local connection close. Other than that, LGTM. 

(I also added some comments previously and deleted them, since I realized I was wrong)

**Comment by rgs1** [comment]
The linux multiarch failure is transient, happened on other PRs too.

File: test/extensions/filters/network/thrift_proxy/conn_manager_test.cc
Comment (wbpcode) [reply test/extensions/filters/network/thrift_proxy/conn_manager_test.cc:1201]: I don't think cx_destroy_local_with_active_rq will increase because now our requests are sent and get responses sequentially. When the connection is disconnected, there will be no requests in pending. But I can try to construct a scenario like this to test that cx_destroy_local_with_active_rq increases.

**Comment by wbpcode** [comment]
@rgs1 
I think `cx_destroy_local_with_active_rq` will never increase. After we get the complete request (in finalizeRequest), we will increase the `accumulated_requests` count without waiting for the response to end. When the count is greater than `max_requests`, the `request_overflow` flag will be set. No new requests will be accepted after that. **After all accepted requests are over, we will actively disconnect.**

The advantage of this is that we don't have to worry about idempotence problems caused by requests that exceed the limit when the connection is disconnected. Because the request exceeds the limit from the beginning, its data will not be forwarded by the thrift proxy.

**Comment by rgs1** [comment]
> @rgs1
> I think `cx_destroy_local_with_active_rq` will never increase. After we get the complete request (in finalizeRequest), we will increase the `accumulated_requests` count without waiting for the response to end. When the count is greater than `max_requests`, the `request_overflow` flag will be set. **No new requests will be accepted after that**. After **all** accepted requests are over, we will actively disconnect.
> 
> The advantage of this is that when the connection is disconnected, we don't have to worry about those requests that exceed the limit being reset and introducing idempotent problems. Because those requests that exceed the limit will not be forwarded by the thrift proxy at all.

Ok makes sense.

**Comment by rgs1** [comment]
You'll want to merge main once #15238 lands.

**Review by rgs1 — APPROVED** [review]
Thanks!

**Review by rgs1 — APPROVED** [review]

**Comment by wbpcode** [comment]
@rgs1 main merged.

**Comment by rgs1** [comment]
@wbpcode LGTM, thanks!

@zuercher over to you for merging.

**Review by zuercher — APPROVED** [review]

**Pull request merged**
